src/transformStateWithClones.js

```diff
@@ -7,7 +7,26 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+  const stateCopy = { ...state };
+
+  for (const action of actions) {
+    if (action.type === 'addProperties') {
+      Object.assign(stateCopy, action.extraData);
+    } else if (action.type === 'removeProperties') {
+      for (const key in action.keysToRemove) {
+        delete stateCopy[key];
+      }
+    } else if (action.type === 'clear') {
+      for (const key of Object.keys(stateCopy)) {
+        delete stateCopy[key];
+      }
+    }
+
+    stateHistory.push(stateCopy);
```

Comment:
Currently, `stateCopy` is being pushed to `stateHistory` without creating a new copy. This means all entries in `stateHistory` will reference the same object. You should create a new copy of `stateCopy` (e.g., using `{ ...stateCopy }`) before pushing it to `stateHistory` to ensure each entry is a separate snapshot of the state.